engine/src/main/java/org/pentaho/di/trans/Trans.java

```diff
@@ -0,0 +1,5804 @@
+      db.connect();
+      db.setCommit( logCommitSize );
+
+      List<LoggingHierarchy> loggingHierarchyList = getLoggingHierarchy();
+      for ( LoggingHierarchy loggingHierarchy : loggingHierarchyList ) {
+        db.writeLogRecord( channelLogTable, LogStatus.START, loggingHierarchy, null );
+      }
+
+      // Also time-out the log records in here...
+      //
+      db.cleanupLogRecords( channelLogTable, getName() );
+    } catch ( Exception e ) {
+      throw new KettleException( BaseMessages.getString( PKG,
+        "Trans.Exception.UnableToWriteLogChannelInformationToLogTable" ), e );
+    } finally {
+      disconnectDb( db );
+    }
+  }
+
+  /**
+   * Writes step information to a step logging table (if one has been configured).
+   *
+   * @throws KettleException if any errors occur during logging
+   */
+  protected void writeStepLogInformation() throws KettleException {
+    Database db = null;
+    StepLogTable stepLogTable = getTransMeta().getStepLogTable();
+    try {
+      db = createDataBase( stepLogTable.getDatabaseMeta() );
+      db.shareVariablesWith( this );
+      db.connect();
+      db.setCommit( logCommitSize );
+
+      for ( StepMetaDataCombi combi : getSteps() ) {
+        db.writeLogRecord( stepLogTable, LogStatus.START, combi, null );
+      }
+
+      db.cleanupLogRecords( stepLogTable, getName() );
+    } catch ( Exception e ) {
+      throw new KettleException( BaseMessages.getString( PKG,
+        "Trans.Exception.UnableToWriteStepInformationToLogTable" ), e );
+    } finally {
+      disconnectDb( db );
+    }
+
+  }
+
+  protected Database createDataBase( DatabaseMeta meta ) {
+    return new Database( this, meta );
+  }
+
+  protected synchronized void writeMetricsInformation() throws KettleException {
+    Preconditions.checkNotNull( log );
+    List<MetricsDuration> metricsList =
+      MetricsUtil.getDuration( log.getLogChannelId(), Metrics.METRIC_PLUGIN_REGISTRY_REGISTER_EXTENSIONS_START );
+    if ( ( log.isDebug() ) && !metricsList.isEmpty() ) {
+      log.logDebug( metricsList.get( 0 ).toString() );
+    }
+
+    metricsList =
+      MetricsUtil.getDuration( log.getLogChannelId(), Metrics.METRIC_PLUGIN_REGISTRY_PLUGIN_REGISTRATION_START );
+    if ( ( log != null ) && ( log.isDebug() ) && !metricsList.isEmpty() ) {
+      log.logDebug( metricsList.get( 0 ).toString() );
+    }
+
+    long total = 0;
+    metricsList =
+      MetricsUtil.getDuration( log.getLogChannelId(), Metrics.METRIC_PLUGIN_REGISTRY_PLUGIN_TYPE_REGISTRATION_START );
+    if ( ( log != null ) && ( log.isDebug() ) && metricsList != null && !metricsList.isEmpty() ) {
+      for ( MetricsDuration duration : metricsList ) {
+        total += duration.getDuration();
+        log.logDebug( "   - " + duration.toString() + "  Total=" + total );
+      }
+    }
+
+    Database db = null;
+    MetricsLogTable metricsLogTable = transMeta.getMetricsLogTable();
+    try {
+      db = new Database( this, metricsLogTable.getDatabaseMeta() );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *db*. The resource is referenced at lines 2675, 2676, 2677, 2687, 2698, 2706, 2711. The resource is closed at line 2722. There are other execution paths that do not contain closure statements, for example, when Database.commit (in method disconnectDb) throws an exception, and an if condition at line 2719 evaluates to true.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).